### PR TITLE
Add pre-commit configuration

### PR DESCRIPTION
**Pull request by Guts**
**In this PR**

- add classical git hooks through [pre-commit](https://pre-commit.com/) to prevent "simple" issues when pushing code
- add basic flake8 git hook, reusing configuration from setup.cfg and selecting most important flags
- add specific section for https://pre-commit.ci/ - recommended because it automates hook updates and quick fixes when it's possible. This requires that a member with write permissions registers the repository on the application.
- add required as extra dependencies into setup.py for development

**Goal**

A little step to improve code reliability.

**Related issues**

Related to #755

### CODE REVIEW

**Comment by Guts** [comment]
ping @jelmer and @ix5

**Review by jelmer — APPROVED** [review]

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        args: [--markdown-linebreak-ext=md]
+
+  - repo: https://github.com/pycqa/flake8
+    rev: 4.0.1
+    hooks:
+      - id: flake8
+        language: python
+        files: ^isso/.*\.py$
+        types: [python]
+        args: ["--config=setup.cfg", "--select=E9,F63,F7,F82"]
```
Comment (jelmer) [inline .pre-commit-config.yaml:30]: isn't this severely limiting what errors we'll get?

Comment (Guts) [reply .pre-commit-config.yaml:30]: It's not to be too frustrating at the time of the commit or a PR.
If you prefer, we can enlarge this selection or apply everything.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+# Specific part for https://pre-commit.ci
```
Comment (ix5) [inline .pre-commit-config.yaml:32]: Am I understanding this correctly that there'll be a little bot auto-committing to this repo? I'd rather avoid this. Proposing changes is fine, but commit rights seem a bit overkill.
Or will this only auto-commit what it thinks are fixes to the branches of PR authors which we can then squash?

**Review by ix5 — COMMENTED** [review]
I like your effort, @Guts. This certainly seems like something that could help catch little slip-ups in a fast-paced development environment.
I've left a few comments in the code.

In general though, I think running linting tools should be up to the discretion of individual committers. We don't want to force them to install _yet another_ framework on their machine (with auto-downloaded hook repos from a to their eyes untrusted Github source).

I'm perfectly fine with _recommending_ running flake8 before committing. flake8 is being run as CI on PRs already.

I'm also a bit confused because `pre-commit` seems to be two things at the same time:
- A framework for running "hooks" from its GitHub hook repository (which needs to be updated regularly) by creating a virtualenv (which I can't find any info about)
- A CI cloud service (free?) which requires push access for its bot to contributors' branches to push what it thinks are trivial fixes

The first one, as I've stated above, should in my opinion reside with individual contributors. We shouldn't force anyone to use it, but having the config for people who might want it seems a good idea.
The second one seems a bit intrusive. It's another unknown actor to trust. For my taste, the whole "sprinkle some packages on it" vibe of nodejs is not something I'd want to bring to this project. But then again, I'd not be entirely opposed to it. Waiting for other maintainers' opinions on this.

Don't get me wrong, I'm very much grateful for your efforts @Guts, but it seems this project will soon be 90% configs for some type of devops service while actual, structural or architectural issues are left hanging. If this helps us write unit tests or invites more contributors, that's a positive outcome. But as they say, adding fancy new things is easy, fixing bugs is hard ;)

**Comment by jelmer** [comment]
> In general though, I think running linting tools should be up to the discretion of individual committers. We don't want to force them to install _yet another_ framework on their machine (with auto-downloaded hook repos from a to their eyes untrusted Github source).
> 
> I'm perfectly fine with _recommending_ running flake8 before committing. flake8 is being run as CI on PRs already.
Hmm, I'd forgotten about that. Adding flake8 to the "dev" dependencies seems reasonable. I'm not sure that the pre-commit config really adds much in that case.

-1 on automatic updates by bot from me as well.

Jelmer

**Pull request closed without merging**